metadata-io/src/main/java/com/linkedin/metadata/entity/DeleteReferencesService.java

```diff
@@ -0,0 +1,177 @@
+package com.linkedin.metadata.entity;
+
+import com.linkedin.data.DataComplex;
+import com.linkedin.data.DataList;
+import com.linkedin.data.DataMap;
+import com.linkedin.data.schema.ArrayDataSchema;
+import com.linkedin.data.schema.DataSchema;
+import com.linkedin.data.schema.PathSpec;
+import com.linkedin.data.schema.RecordDataSchema;
+import com.linkedin.data.template.RecordTemplate;
+import com.linkedin.entity.Aspect;
+import java.util.List;
+import java.util.ListIterator;
+import lombok.extern.slf4j.Slf4j;
+
+
+/**
+ * Utility class that encapsulates the logic on how to modify a {@link RecordTemplate} in place to remove a single value
+ * following a concrete set of rules.
+ *
+ * It does this by a recursive tree traversal method, based on an aspect path provided for a given aspect.
+ * This so that we don't have to scan the entire aspect for the value to be removed and then figure out how to apply
+ * logical rules based on upstream optionality definitions.
+ *
+ * For more information see {@link #getAspectWithReferenceRemoved(String, RecordTemplate, DataSchema, PathSpec)}
+ */
+@Slf4j
+public class DeleteReferencesService {
+
+  private DeleteReferencesService() { }
+
+  /**
+   *
+   * Utility method that provides utility methods to remove fields from a given aspect based on it's aspect spec that
+   * follows the following logic:
+   *
+   * 1. If field is optional and not part of an array → remove the field.
+   * 2. If is a field that is part of an array (has an `*` in the path spec)
+   *  → go up to the nearest array and remove the element.
+   *  Extra → If array only has 1 element which is being deleted→ optional rules (if optional set null, otherwise delete)
+   * 3. If field is non-optional and does not belong to an array delete if and only if aspect becomes empty.
+   *
+   * @param value       Value to be removed from Aspect
+   * @param aspect      Aspect in which the value property exists
+   * @param schema      {@link DataSchema} of the aspect being processed
+   * @param aspectPath  Path within the aspect to where the value can be found.
+   * @return  An updated version of the provided aspect without the provided value.
+   */
+  public static Aspect getAspectWithReferenceRemoved(String value, RecordTemplate aspect, DataSchema schema, PathSpec aspectPath) {
+    try {
+      final DataMap copy = aspect.copy().data();
+      final DataComplex newValue = removeValueBasedOnPath(value, schema, copy, aspectPath.getPathComponents(), 0);
+      return new Aspect((DataMap) newValue);
+    } catch (CloneNotSupportedException e) {
+      return new Aspect();
```

Comment:
What is this return doing? What are the implications?

Reply:
This exception is never triggered, but we still have to catch it. Not really sure why PDL has this exception thrown on all clones

Reply:
should we throw or log or something? Feels a bit insidious to just return an empty aspect object